azkaban-common/src/main/java/azkaban/imagemgmt/daos/RampRuleDaoImpl.java

```diff
@@ -147,6 +178,72 @@ public void deleteRampRule(final String ruleName) {
 
   }
 
+  /**
+   * Map FlowId (project.flow) with denying information (denyMode or denyVersion) and insert them into DB.
+   * Deduplication performed in same rule with same flow to avoid DB inflation.
+   *
+   * @param flowIds
+   * @param ruleName
+   * @return int - id of the DB entry
+   */
+  @Override
+  public int addFlowDenyInfo(final List<ProjectFlow> flowIds, final String ruleName) {
+    // query ramp_rules to get ramp rule information
+    // based on ramp rule to generate mappings with flowId as key
+    final SQLTransaction<Long> fetchRampRuleAndUpdateDenyList = transOperator -> {
+      List<ImageRampRule> imageRampRuleQueriedResults =
+          transOperator.query(FetchRampRuleHandler.FETCH_RAMP_RULE_BY_ID, new FetchRampRuleHandler(), ruleName);
+      if (imageRampRuleQueriedResults.isEmpty()) {
+        LOG.error("fail to find the existing ruleName: " + ruleName);
+        throw new ImageMgmtDaoException("ramp rule not found with ruleName: " + ruleName);
+      }
+      ImageRampRule imageRampRule = imageRampRuleQueriedResults.get(0);
+      // insert HP flows: <flowId, denyMode.ALL, ruleName>
+      if (imageRampRule.isHPRule()) {
+        LOG.info("handling add flows for HP Flow Rule: " + ruleName);
+        for (ProjectFlow flowId : flowIds) {
+          // avoid duplicate insert HP flows, but allow different rule defines same hp flow in case deleted case
+          if (transOperator.query(
+                FetchFlowDenyListHandler.FETCH_FLOW_DENY_LIST_BY_FLOW_ID_AND_DENY_MODE,
+                new FetchFlowDenyListHandler(), flowId.toString(), DenyMode.ALL, ruleName)
+              .isEmpty()) {
+            transOperator.update(INSERT_HP_DENY_LIST, flowId.toString(), DenyMode.ALL.name(), ruleName);
+          }
+        }
+      } else {
+        // insert normal flows: <flowId, denyMode.PARTIAL, ruleName>
+        String denyVersion = String.join(":", imageRampRule.getImageName(), imageRampRule.getImageVersion());
+        LOG.info("handling add flows for normal Flow Rule: " + ruleName + " with denyVersion: " + denyVersion);
+        for (ProjectFlow flowId : flowIds) {
+          // avoid duplicate insert, but allow different rule defines same flow with same version
+          if (transOperator.query(
+                  FetchFlowDenyListHandler.FETCH_FLOW_DENY_LIST_BY_FLOW_ID_AND_DENY_VERSION,
+                  new FetchFlowDenyListHandler(), flowId.toString(), denyVersion, ruleName)
+              .isEmpty()) {
+            transOperator.update(INSERT_FLOW_DENY_LIST, flowId.toString(), DenyMode.PARTIAL.name(), denyVersion, ruleName);
+          }
+        }
+      }
```

Comment:
for large code blocks please use comments like "end if" , "end transOperator" etc.

Reply:
added